Add siteless checkout thank you page (and post-checkout redirect).

elliottprogrammer opened the pull request:
#### Changes proposed in this Pull Request

* Adds a new route & controller for the siteless thank-you page.
* Sets the post-checkout redirect (the Thank You page url) to: `/checkout/jetpack/thank-you/no-site/:product?ticketId=321312`
* Adds  new `JetpackSitelessCheckoutThankYou` component & scss styles for rendering the post checkout thank-you page (for _siteless_ purchases).

Still TODO:
- On submitting the website address input, we still need to dispatch a post request to WPCOM with the site address and the and the user email address to some endpoint so we can append the site address(url) to the ZendDesk ticket (which will be created on the wpcom backend after purchase is completed).
- On successful response(of above request), we need to redirect to a "schedule 15min Happiness support" page. (Calendly?)
- Figure out/create the correct link (Calendly link?) for the "schedule a 15 min call now" link. (currently the link is incorrect and not working correctly)
- Figure out how to pass the Zendesk ticket ID (created on the wpcom backend after transaction completed) back to the Calypso checkout thank-you page.

Asana task: 1200479326344990-as-1200481482484418

#### Testing instructions

1. Patch your WPCOM sandbox with D63168-code. (we need this to be able to complete the _site-less_ checkout)
2. Add `define( 'USE_STORE_SANDBOX', true );` to `wp-content/mu-plugins/0-sandbox.php`
3. Sandbox `public-api.wordpress.com`.
4. Download this PR.
5. Edit the `config/development.json` file and enable the `jetpack/siteless-checkout` feature flag.
6. Start Calypso with `yarn start`. You need to do this since this PR introduces a new feature flag.
7. Visit http://calypso.localhost:3000/checkout/jetpack/jetpack_scan on a window where you are **not** logged-in to WordPress.com.
8. Complete checkout (using an email address that is **not** already associated with a WordPress.com account).
9. Make sure you are redirected to the siteless Thank You page. ( `http://calypso.localhost:3000/checkout/jetpack/thank-you/no-site/jetpack_scan`).
10. Verify the Thank You page looks like the screenshot below and like the figma design link, located in the Asana task(link above).
11. Open the browser console and enter: `localStorage.debug='calypso:analytics*'`
12. Reload the page and check the Tracks events:
    - Verify the Page View event is firing on page load.
    - Verify the `calypso_siteless_checkout_install_instructions_link_clicked` event fires when clicking the, "instructions we put together here" link. Verify the `product_slug` property  and value. Verify that the link opens `https://jetpack.com/support/getting-started-with-jetpack/` in a new tab.
    - Verify the `calypso_siteless_checkout_happiness_link_clicked` event fires when clicking the "schedule a 15 min call now" link. Verify the `product_slug` property and value.
13. Verify the "Continue" button is disabled until something is entered into the input field.
14. In the Website address field, enter a site address and click "Continue" button.  Verify the `calypso_siteless_checkout_submit_website_address` event fires. Verify `product_slug`, and `site_url` properties and values.

**Run unit test:**
`yarn test-client client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js`
Verify all tests pass.

#### Screenshot

<img width="1491" alt="Screenshot on 2021-07-02 at 09-34-23" src="https://user-images.githubusercontent.com/11078128/124304774-3345eb00-db32-11eb-932f-a2de4c5a3e50.png">

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

elliottprogrammer committed:
Add siteless checkout thank you page (and post-purchase redirect).

sirbrillig commented on client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:
```diff
@@ -125,14 +125,27 @@ export default function getThankYouPageUrl( {
-		return `/checkout/jetpack/thank-you/${ siteSlug }/${ productSlug ?? 'no_product' }`;
+		// TODO: Get the Zendesk ticket ID & pass it along with the thankYouPageUrl (`ticketId` query param).
+		// Just using the `pendingOrReceiptId` value for now as an example.
+		return `/checkout/jetpack/thank-you/no-site/${
```
Could you add a test for this in https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js, please?

elliottprogrammer replied:
Oops. Oh yes, for sure! - Done. 👍

elliottprogrammer committed:
Fix loading state placeholder & other minor corrections.

elliottprogrammer committed:
Add unit test for composite-checkout-thank-you.

rcanepa commented on client/my-sites/checkout/checkout-thank-you/style.scss:
```diff
@@ -675,6 +675,151 @@
+				max-width: 456px;
+			}
+		}
+		// TODO min width
+		// TODO primary button color
```
Are these `TODO` comments relevant?

elliottprogrammer replied:
No, they're not. Thank you, I'll remove them.

rcanepa commented on client/my-sites/checkout/checkout-thank-you/jetpack-checkout-siteless-thank-you.tsx:
```diff
@@ -0,0 +1,211 @@
+												onClick={ () =>
+													dispatch(
+														recordTracksEvent( 'calypso_siteless_checkout_happiness_link_clicked', {
+															zendeskTicketId,
+															productSlug,
```
These property names need to be converted to snake case.

rcanepa commented on client/my-sites/checkout/checkout-thank-you/jetpack-checkout-siteless-thank-you.tsx:
```diff
@@ -0,0 +1,211 @@
+															recordTracksEvent(
+																'calypso_siteless_checkout_install_instructions_link_clicked',
+																{
+																	zendeskTicketId,
+																	productSlug,
```
These property names need to be converted to snake case.

rcanepa commented on client/my-sites/checkout/checkout-thank-you/jetpack-checkout-siteless-thank-you.tsx:
```diff
@@ -0,0 +1,211 @@
+		if ( siteUrl ) {
+			dispatch(
+				recordTracksEvent( 'calypso_siteless_checkout_submit_website_address', {
+					zendeskTicketId,
+					productSlug,
+					siteUrl,
```
These property names need to be converted to snake case.

rcanepa commented:
It works great and looks awesome! I added Jetpack Design to the list of reviewers because they will provide better feedback than us in that regard.

We changed the approach, and now we will not share the ticket ID with the client (please read p1HpG7-cfR-p2#comment-47480 to get up to speed). Instead, we will send the website URL to the backend and update the ticket based on a unique tag and the user email address (that will be our key to finding it). So, feel free to delete all references to the ticket ID.

rcanepa commented:
I noticed in the console that Tracks is complaining because the name of the properties does not follow our rules.

![image](https://user-images.githubusercontent.com/3418513/124524951-33144c80-ddcb-11eb-8030-9c34adc86ce5.png)

elliottprogrammer committed:
Remove irrelevant TODO comments in scss.

elliottprogrammer committed:
Make the Tracks property names camel case.

elliottprogrammer committed:
Remove references to the zendeskTicketId.

elliottprogrammer commented:
Awesome @rcanepa, thanks for the thorough review. 👍

> I added Jetpack Design to the list of reviewers because they will provide better feedback than us in that regard.

Oh yes, I forgot to add them, thank you. 

> We changed the approach, and now we will not share the ticket ID with the client.

Awesome, thanks for the info on that.

I removed the irrelevant TODO comments, made the Tracks properties camel case, and removed references to the Zendesk ticket id. 👍

rcanepa reviewed (approved):
Thank you for addressing the comments I left before. 

It looks good to me now. We can merge this work and make sure we have cards on our board to tackle the pending parts: the website URL and Calendly integrations.

elliottprogrammer commented:
@keoshi, could you take a look at this PR and give a design review?

rcanepa commented:
> @keoshi, could you take a look at this PR and give a design review?

@keoshi is AFK so let's get this merged. We can fix design issues on a follow-up PR when he is back.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6157665

Thank you @elliottprogrammer for including a screenshot in the description! This is really helpful for our translators.

keoshi commented:
@elliottprogrammer @rcanepa Should I create a new PR to fix the design issues?

elliottprogrammer commented:
> Should I create a new PR to fix the design issues?

Yes sure! If you don't mind.  Thanks @keoshi !!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6157665).
